### PR TITLE
fix(models): pad attention mask by remaining length instead of target length (#14699)

### LINKED ISSUE #14699: Fix skipped `test_model_xattn_padding` test by updating mask padding logic
### **Title:**
Fix skipped `test_model_xattn_padding` test by updating mask padding logic

### **Description:**
Currently, there is a skipped test in `tests/models/unets/test_models_unet_2d_condition.py` regarding cross-attention mask padding (`test_model_xattn_padding`).

The test is currently skipped with the following reason:
> *"we currently pad mask by target_length tokens (what unclip needs), whereas stable-diffusion's cross-attn needs to instead pad by remaining_length."*

This is reflected by a dangling `TODO` in `src/diffusers/models/attention_processor.py` inside `prepare_attention_mask` (around line 740):

```python
# TODO: for pipelines such as stable-diffusion, padding cross-attn mask:
#       we want to instead pad by (0, remaining_length), where remaining_length is:
#       remaining_length: int = target_length - current_length
# TODO: re-enable tests/models/test_models_unet_2d_condition.py#test_model_xattn_padding
attention_mask = F.pad(attention_mask, (0, target_length), value=0.0)
```

### **Expected Behavior / Proposed Solution:**
1. Update `prepare_attention_mask` in `attention_processor.py` (and any other relevant `attention.py` files) to pad the mask by `remaining_length = target_length - current_length`. 
2. Ensure the MPS `torch.zeros()` hack is also updated to pad by `remaining_length` rather than `target_length`.
3. Verify that these changes do not break existing `unclip` tests/pipelines.
4. Remove `@pytest.mark.skip` from `test_model_xattn_padding` in `tests/models/unets/test_models_unet_2d_condition.py` to re-enable it.

### **Relevant Files:**
- `src/diffusers/models/attention_processor.py`
- `tests/models/unets/test_models_unet_2d_condition.py`

@sayakpaul  If this is indeed technical debt that needs fixing, I would be happy to put together a PR to update the logic and re-enable the test myself!

### PR DESCRIPTION
**Pull request by dlowzzxx**
﻿Fixes #14699

### Problem Description
Previously in `prepare_attention_mask` (in both `src/diffusers/models/attention_processor.py` and `src/diffusers/models/attention.py`), when padding an attention mask whose length was smaller than `target_length`, `F.pad(attention_mask, (0, target_length), value=0.0)` was used.

Because `torch.nn.functional.pad(tensor, (0, n))` appends `n` padding elements to the trailing dimension, passing `target_length` resulted in a post-padding length of `current_length + target_length` rather than `target_length`. The same issue occurred on MPS where `padding_shape` used `target_length` instead of the difference between `target_length` and `current_length`.

Due to this bug, `test_model_xattn_padding` in `tests/models/unets/test_models_unet_2d_condition.py` was skipped with a TODO noting:
> *"we currently pad mask by target_length tokens (what unclip needs), whereas stable-diffusion's cross-attn needs to instead pad by remaining_length."*

### Proposed Changes
1. Updated `prepare_attention_mask` in both `src/diffusers/models/attention_processor.py` and `src/diffusers/models/attention.py`:
   - Compute `remaining_length = target_length - current_length`.
   - When `remaining_length > 0`, pad by `remaining_length`:
     - On MPS, construct `padding_shape = (attention_mask.shape[0], attention_mask.shape[1], remaining_length)`.
     - Otherwise, call `F.pad(attention_mask, (0, remaining_length), value=0.0)`.
2. Unskipped `test_model_xattn_padding` in `tests/models/unets/test_models_unet_2d_condition.py` and aligned `block_out_channels = (16, 32)` to ensure divisibility with `attention_head_dim = (8, 16)`.
3. Cleaned up unused imports and formatted code with `ruff`.

### Verification & Reproduction
Ran the unskipped test directly:
```bash
pytest tests/models/unets/test_models_unet_2d_condition.py -k test_model_xattn_padding -v
```
Output:
```
============================= test session starts =============================
platform win32 -- Python 3.12.10, pytest-9.1.1, pluggy-1.6.0
collected 164 items / 163 deselected / 1 selected

tests/models/unets/test_models_unet_2d_condition.py::TestUNet2DCondition::test_model_xattn_padding PASSED [100%]
================ 1 passed, 163 deselected, 5 warnings in 0.97s ================
```

Also verified MPS mock and boundary cases (mask is None, `current_length == target_length`, `current_length > target_length`). All checks and `ruff check`/`ruff format` passed cleanly.

### CODE REVIEW

**Review by siddhant-shahhh — APPROVED** [review]
This is a great catch and a highly satisfying fix. The previous implementation fundamentally misapplied the `F.pad` API by treating the padding parameter as a target dimension size rather than an absolute addition count, which silently bloated the attention mask tensors.

### Code Observations
1. **Mathematical Accuracy:** Calculating `remaining_length = target_length - current_length` properly satisfies the PyTorch padding tuple requirements.
2. **MPS Compatibility:** Ensuring the custom MPS padding block also adopts `remaining_length` keeps the device execution paths symmetrically aligned.
3. **Test Restoration:** It is always a huge win to successfully un-skip a lingering test. Modifying the block out channels to cleanly divide by the attention head dims makes the test completely deterministic.


<summary><b>Review Environment</b></summary>

- **OS:** Windows (64-bit)
- **Python:** CPython 3.12.10
- **Verification:** Pulled branch `fix/xattn-mask-padding-remaining-length-14699`, inspected `attention.py` and `attention_processor.py`, and successfully ran `pytest tests/models/unets/test_models_unet_2d_condition.py -k test_model_xattn_padding` locally.



Clean logic, thorough testing, and excellent documentation of the fix. LGTM!
